spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/listener/AbstractHazelcastEventListener.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+package org.springframework.integration.hazelcast.listener;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EventObject;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.MultiMap;
+
+import org.springframework.integration.hazelcast.common.CacheListeningPolicyType;
+import org.springframework.integration.hazelcast.context.ApplicationContextStartEventHandler;
+import org.springframework.integration.hazelcast.inbound.HazelcastMessageProducer;
+import org.springframework.messaging.support.MessageBuilder;
+
+/**
+ * This is abstract Hazelcast event listener for entry/item/message events.
+ * 
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ *
+ */
+public abstract class AbstractHazelcastEventListener {
+
+	private HazelcastMessageProducer hazelcastInboundChannelAdapter;
+
+	public AbstractHazelcastEventListener(HazelcastMessageProducer hazelcastInboundChannelAdapter) {
+		this.hazelcastInboundChannelAdapter = hazelcastInboundChannelAdapter;
+	}
+
+	public HazelcastMessageProducer getHazelcastInboundChannelAdapter() {
+		return hazelcastInboundChannelAdapter;
+	}
+
+	protected abstract void processEvent(EventObject event);
+
+	protected Set<SocketAddress> getLocalSocketAddresses() {
+		Set<SocketAddress> localSocketAddressesSet = new HashSet<>();
+		for (HazelcastInstance hazelcastInstance : Hazelcast.getAllHazelcastInstances()) {
+			localSocketAddressesSet.add(hazelcastInstance.getLocalEndpoint().getSocketAddress());
+		}
+
+		return localSocketAddressesSet;
+	}
+
+	protected void sendMessage(EventObject event, InetSocketAddress socketAddress,
+			CacheListeningPolicyType cacheListeningPolicyType) {
+		
+		if (CacheListeningPolicyType.ALL == cacheListeningPolicyType
+				|| isEventAcceptable(socketAddress)) {
+
+			getHazelcastInboundChannelAdapter().sendMessage(
+					MessageBuilder.withPayload(event).build());
+
+		}
+	}
+
+	private boolean isEventAcceptable(InetSocketAddress socketAddress) {
+		Set<SocketAddress> localSocketAddressesSet = getLocalSocketAddresses();
+		if ((!localSocketAddressesSet.isEmpty())
+				&& (localSocketAddressesSet.contains(socketAddress) || 
+						isEventAcceptable(localSocketAddressesSet, socketAddress))) {
+			return true;
+		}
+
+		return false;
+	}
+
+	private boolean isEventAcceptable(Set<SocketAddress> localSocketAddressesSet,
+			InetSocketAddress socketAddressOfEvent) {
+		HazelcastInstance hazelcastInstance = Hazelcast.getAllHazelcastInstances().iterator().next();
+		MultiMap<SocketAddress, SocketAddress> configMultiMap = hazelcastInstance
+				.getMultiMap(ApplicationContextStartEventHandler.HZ_INTERNAL_CONFIGURATION_MULTI_MAP);
+		if (configMultiMap.size() > 0
+				&& !configMultiMap.values().contains(socketAddressOfEvent)
+				&& localSocketAddressesSet.contains(configMultiMap.keySet().iterator().next())) {
```

Comment:
@erenavsarogullari , whould you mind explaining the logic here.

I see a vulnerability here: the event may be initiated non-SI application, but by a member of our cluster. Hence our event can be handled only by one node - the first one in the in this our distributed Map. And unfortunately, always only by the same node, because we use only the first member.

Really there is no something like round-robing in Hazelcast?
I mean, if we handle an event only once, let's distribute that's on the cluster! The first event by first node, the second - by second and so on.
I think we can use distributed `IAtomicLong` to determine the `index` for the node.